### PR TITLE
Gitlab #270 : webadmin address/groups endpoints

### PR DESCRIPTION
**Pull request by mbaechler**
What is left to do : 
- **DONE** fixing all other webadmin endpoints to actually return some JSON
- **DONE** writing some integration tests to check that the groups endpoint is at the expected path and answering
- **DONE** maybe writing some tests for unexpected exceptions using mocks
- **DONE** check that it works in cassandra project
- **DONE** verify what is served as swagger documentation

Added by @chibenwa : 
 - **DONE** Also factorize SMTP mail address
 - **DONE** Write website DOC

### CODE REVIEW

**Comment by chibenwa** [comment]
FYI https://github.com/perwendel/spark/issues/911default content type in Spark Java

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/UsersRoutesTest.java [hunk 1]
```diff
@@ -155,13 +165,20 @@ public void postShouldReturnRequireNonNullPassword() {
...
+            List<Map<String, String>> users =
+                when()
+                    .get()
+                .then()
+                    .statusCode(200)
+                    .contentType(ContentType.JSON)
+                    .extract()
+                    .body()
+                    .jsonPath()
+                    .getList(".flatten()");
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/UsersRoutesTest.java]: why do you need this ugly `flatten()` ?

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRouteTest.java [hunk 2]
```diff
@@ -308,6 +316,150 @@ public void deleteRequiresTwoPathParams() {
...
+
+        @Test
+        public void putShouldReturnErrorWhenRecipientRewriteTableExceptionIsThrown() throws Exception {
+            doThrow(RecipientRewriteTableException.class)
+                .when(memoryRecipientRewriteTable)
+                .addAddressMapping(anyString(), anyString(), anyString());
+
+            when()
+                .put(GROUP1 + SEPARATOR + GROUP2)
+                .then()
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRouteTest.java]: indent issue

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/UsersRoutesTest.java (hunk 1 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/UsersRoutesTest.java]: Not needed. It is a typo.

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRouteTest.java (hunk 2 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRouteTest.java]: Ok

File: server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java [hunk 3]
```diff
@@ -23,6 +23,8 @@
 
 public interface Mapping {
 
+    String getAddress();
+
```
Comment (rouazana) [inline server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java]: not related to this commit

File: server/data/data-library/src/main/java/org/apache/james/rrt/lib/MappingImpl.java [hunk 4]
```diff
@@ -33,6 +33,7 @@
     private static final long serialVersionUID = 1L;
 
     private static final String ADDRESS_PREFIX = "";
+    private boolean address;
```
Comment (rouazana) [inline server/data/data-library/src/main/java/org/apache/james/rrt/lib/MappingImpl.java]: all the changes of this file are unrelated too

File: server/data/data-library/src/test/java/org/apache/james/rrt/lib/MappingImplTest.java [hunk 5]
```diff
@@ -21,6 +21,7 @@
 package org.apache.james.rrt.lib;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
```
Comment (rouazana) [inline server/data/data-library/src/test/java/org/apache/james/rrt/lib/MappingImplTest.java]: idem

File: server/protocols/protocols-lmtp/src/main/java/org/apache/james/lmtpserver/DataLineLMTPHandler.java [hunk 6]
```diff
@@ -53,12 +54,12 @@ protected Response processExtensions(SMTPSession session, Mail mail) {
         // build a wrapper around the Mail
         final ReadOnlyMailEnvelope env = new ReadOnlyMailEnvelope(mail);
 
-        for (org.apache.mailet.MailAddress recipient : mail.getRecipients()) {
+        for (org.apache.james.core.MailAddress recipient : mail.getRecipients()) {
```
Comment (rouazana) [inline server/protocols/protocols-lmtp/src/main/java/org/apache/james/lmtpserver/DataLineLMTPHandler.java]: really need the fqdn?

File: server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/DataLineJamesMessageHookHandler.java [hunk 7]
```diff
@@ -102,7 +100,7 @@ public Response onLine(SMTPSession session, ByteBuffer lineByteBuffer, LineHandl
                 List<MailAddress> recipientCollection = (List<MailAddress>) session.getAttachment(SMTPSession.RCPT_LIST, State.Transaction);
                 MailAddress mailAddress = (MailAddress) session.getAttachment(SMTPSession.SENDER, State.Transaction);
 
-                List<org.apache.mailet.MailAddress> rcpts = new ArrayList<>();
+                List<org.apache.james.core.MailAddress> rcpts = new ArrayList<>();
```
Comment (rouazana) [inline server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/DataLineJamesMessageHookHandler.java]: idem in this file?

File: server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/model/ProtocolMailAddressAdapter.java [hunk 8]
```diff
@@ -18,13 +18,15 @@
...
+
 import org.apache.james.protocols.smtp.MailAddress;
-import org.apache.james.protocols.smtp.MailAddressException;
 
+@Deprecated
 public class ProtocolMailAddressAdapter extends MailAddress {
     
-    public ProtocolMailAddressAdapter(org.apache.mailet.MailAddress mailAddress) throws MailAddressException {
-        super(mailAddress.toString());
+    public ProtocolMailAddressAdapter(org.apache.james.core.MailAddress mailAddress) throws AddressException {
```
Comment (rouazana) [inline server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/model/ProtocolMailAddressAdapter.java]: idem

File: src/site/xdoc/server/dev-extend-mailet.xml [hunk 9]
```diff
@@ -82,24 +82,25 @@
     </p>
 
 <source>
-package com.test;
+    package com.test;
```
Comment (rouazana) [inline src/site/xdoc/server/dev-extend-mailet.xml]: this file changeset is weird, no?

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+        @ApiResponse(code = 200, message = "OK", response = List.class),
+        @ApiResponse(code = 500, message = "Internal server error - Something went bad on the server side.")
+    })
+    private Set<String> listGroups(Request request, Response response) throws RecipientRewriteTableException {
+        return Optional.ofNullable(recipientRewriteTable.getAllMappings())
+            .map(mappings ->
+                mappings.entrySet().stream()
+                    .filter(e -> e.getValue().contains(Mapping.Type.Address))
+                    .map(Map.Entry::getKey)
+                    .collect(Guavate.toImmutableSortedSet()))
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java]: no need to explicitly sort on alphabetical ordering?

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRoutesTest.java [hunk 11]
```diff
@@ -0,0 +1,465 @@
...
+    }
+
+    public class NormalBehaviour {
+
+        @Before
+        public void setUp() throws Exception {
+            MemoryRecipientRewriteTable memoryRecipientRewriteTable = new MemoryRecipientRewriteTable();
+            memoryRecipientRewriteTable.addErrorMapping("error", DOMAIN, "disabled");
+            memoryRecipientRewriteTable.addRegexMapping("regex", DOMAIN, ".*@b\\.com");
+            memoryRecipientRewriteTable.addAliasDomainMapping("alias", DOMAIN);
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRoutesTest.java]: you never use these mappings. I understand that you put them to check that they does not interfere, but maybe you should create them in an other test case, so that you have: NormalBehaviour, ErrorBehaviour, NormalBehaviourWithAlreadyDefinedRules

File: server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/CassandraMigrationRoutes.java [hunk 12]
```diff
@@ -34,7 +34,6 @@
 
 public class CassandraMigrationRoutes implements Routes {
 
-
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/CassandraMigrationRoutes.java]: do we need to warn admin team about this commit?

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java [hunk 13]
```diff
@@ -261,6 +261,21 @@ public void postShouldDoMigrationAndUpdateToTheLatestVersion() throws Exception
     }
 
     @Test
+    public void test() throws Exception {
```
Comment (rouazana) [inline server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: test name

File: src/site/markdown/server/manage-webadmin.md [hunk 14]
```diff
@@ -425,3 +425,70 @@ Note that several calls to this endpoint will be run in a sequential pattern.
...
+ - 404: The group does not exist
+ - 500: Internal error
+
+### Adding a group member
+
+```
+curl -XPUT http://ip:port/address/groups/group@domain.com/member@domain.com
+```
+
+Will add member@domain.com to group@domain.com
```
Comment (rouazana) [inline src/site/markdown/server/manage-webadmin.md]: Should explain that it will create the group if it does not exist

File: src/site/markdown/server/manage-webadmin.md [hunk 15]
```diff
@@ -425,3 +425,70 @@ Note that several calls to this endpoint will be run in a sequential pattern.
...
+ - 400: Group structure or member is not valid
+ - 500: Internal error
+
+### Removing a group member
+
+```
+curl -XPUT http://ip:port/address/groups/group@domain.com/member@domain.com
+```
+
+Will remove member@domain.com from group@domain.com
```
Comment (rouazana) [inline src/site/markdown/server/manage-webadmin.md]: Should explain that it will remove the group if empty

File: server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/CassandraMigrationRoutes.java (hunk 12 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/CassandraMigrationRoutes.java]: They are not supposed to know about Cassandra migration management (yet)

File: server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java (hunk 3 above)
Comment (mbaechler) [reply server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java]: true

Comment (mbaechler) [reply server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java]: painful but done

File: server/data/data-library/src/main/java/org/apache/james/rrt/lib/MappingImpl.java (hunk 4 above)
Comment (mbaechler) [reply server/data/data-library/src/main/java/org/apache/james/rrt/lib/MappingImpl.java]: fixed

File: server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java (hunk 3 above)
Comment (mbaechler) [reply server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java]: fixed

File: server/data/data-library/src/test/java/org/apache/james/rrt/lib/MappingImplTest.java (hunk 5 above)
Comment (mbaechler) [reply server/data/data-library/src/test/java/org/apache/james/rrt/lib/MappingImplTest.java]: fixed

File: server/protocols/protocols-lmtp/src/main/java/org/apache/james/lmtpserver/DataLineLMTPHandler.java (hunk 6 above)
Comment (mbaechler) [reply server/protocols/protocols-lmtp/src/main/java/org/apache/james/lmtpserver/DataLineLMTPHandler.java]: no

File: server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/DataLineJamesMessageHookHandler.java (hunk 7 above)
Comment (mbaechler) [reply server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/DataLineJamesMessageHookHandler.java]: fixed

File: server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/model/ProtocolMailAddressAdapter.java (hunk 8 above)
Comment (mbaechler) [reply server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/model/ProtocolMailAddressAdapter.java]: fixed

File: src/site/xdoc/server/dev-extend-mailet.xml (hunk 9 above)
Comment (mbaechler) [reply src/site/xdoc/server/dev-extend-mailet.xml]: intellij automagic ... fixed

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java (hunk 10 above)
Comment (mbaechler) [reply server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java]: String + naturalOrdering looks fine to me

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java (hunk 13 above)
Comment (mbaechler) [reply server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: ok

File: src/site/markdown/server/manage-webadmin.md (hunk 15 above)
Comment (mbaechler) [reply src/site/markdown/server/manage-webadmin.md]: yes, and there's a bug on the curl -XPUT, it should be curl -XDELETE

File: src/site/markdown/server/manage-webadmin.md (hunk 14 above)
Comment (mbaechler) [reply src/site/markdown/server/manage-webadmin.md]: ok

File: server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/CassandraMigrationRoutes.java (hunk 12 above)
Comment (rouazana) [reply server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/CassandraMigrationRoutes.java]: I mean changing default output fro html+json to plain json

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRoutesTest.java (hunk 11 above)
Comment (mbaechler) [reply server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRoutesTest.java]: how would you do it ? Basically, I would like all the tests from NormalBehaviour to be replayed in NormalBehaviourWithAlreadyDefinedRules

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java (hunk 10 above)
Comment (rouazana) [reply server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java]: then you mean lexicographic ordering I guess

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRoutesTest.java (hunk 11 above)
Comment (rouazana) [reply server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRoutesTest.java]: I know you don't like it, but the simpler way is to have one class extending the other one, or nicer having them extending a common abstract class

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java (hunk 10 above)
Comment (mbaechler) [reply server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java]: it's not obvious to me what the difference is. Anyway, I should not have wrote that, it's your fault, because ordering is related to l10n and we don't have any clue how to do that for now.

Comment (rouazana) [reply server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java]: in lexicographic, Z < a
anyway I don't care, I just want mappings to be returned in a consistent order. So just remove "with alphabetic order" and I'm happy

File: server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/CassandraMigrationRoutes.java (hunk 12 above)
Comment (mbaechler) [reply server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/CassandraMigrationRoutes.java]: it's probably what they expect and didn't noticed the bug yet. But, yes, we'll tell them.

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java (hunk 10 above)
Comment (mbaechler) [reply server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java]: ok

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRoutesTest.java (hunk 11 above)
Comment (mbaechler) [reply server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRoutesTest.java]: ugly inheritance done

**Comment by quynhn** [comment]
>1) No implementation for java.lang.Long annotated with @com.google.inject.name.Named(value=tokenExpirationInMs) was bound.
  while locating java.lang.Long annotated with @com.google.inject.name.Named(value=tokenExpirationInMs)
    for parameter 1 at org.apache.james.jmap.cassandra.access.CassandraAccessTokenDAO.<init>(CassandraAccessTokenDAO.java:56)
  while locating org.apache.james.jmap.cassandra.access.CassandraAccessTokenDAO
    for parameter 0 at org.apache.james.jmap.cassandra.access.CassandraAccessTokenRepository.<init>(CassandraAccessTokenRepository.java:37)
  at org.apache.james.modules.data.CassandraJmapModule.configure(CassandraJmapModule.java:42) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.data.CassandraJmapModule)

2) No implementation for org.apache.james.modules.server.CamelMailetContainerModule$DefaultProcessorsConfigurationSupplier was bound.
  while locating org.apache.james.modules.server.CamelMailetContainerModule$DefaultProcessorsConfigurationSupplier
    for parameter 6 at org.apache.james.modules.server.CamelMailetContainerModule$MailetModuleConfigurationPerformer.<init>(CamelMailetContainerModule.java:117)
  at org.apache.james.modules.server.CamelMailetContainerModule.configure(CamelMailetContainerModule.java:79) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.MailetProcessingModule -> org.apache.james.modules.server.CamelMailetContainerModule)

3) No implementation for org.apache.james.queue.api.MailQueueItemDecoratorFactory was bound.
  while locating org.apache.james.queue.api.MailQueueItemDecoratorFactory
    for parameter 1 at org.apache.james.queue.activemq.ActiveMQMailQueueFactory.<init>(ActiveMQMailQueueFactory.java:44)
  at org.apache.james.modules.server.ActiveMQQueueModule.createActiveMailQueueFactory(ActiveMQQueueModule.java:52) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.server.ActiveMQQueueModule)

4) No implementation for org.apache.james.util.date.ZonedDateTimeProvider was bound.
  while locating org.apache.james.util.date.ZonedDateTimeProvider
    for parameter 0 at org.apache.james.jmap.cassandra.vacation.CassandraNotificationRegistry.<init>(CassandraNotificationRegistry.java:46)
  at org.apache.james.modules.data.CassandraJmapModule.configure(CassandraJmapModule.java:48) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.data.CassandraJmapModule)

**Comment by quynhn** [comment]
Could we support add a list of user to group instead of one by one?
While writing IT I have to repeat a lot of lines to add multiple users but for further end user prefers adding multiple users IMO

**Comment by chibenwa** [comment]
> Could we support add a list of user to group instead of one by one?

No due to concurrency issues. CF discussions on chat (you may need to scroll in the history...)

**Comment by chibenwa** [comment]
If you post a stacktrace in a PR @quynhn , please:

 - use code to display it.
 - Show the test name failing and project as well.

Because as is, I still need to open CI to get context about the error.

**Comment by chibenwa** [comment]
```
[ERROR] Errors: 
[ERROR]   TestDiscardProtocol.testRecord:78 »  Timed out wait for be notified that read ...
[INFO] 
[ERROR] Tests run: 21, Failures: 0, Errors: 1, Skipped: 0
```

test this please

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java [hunk 16]
```diff
@@ -272,7 +272,7 @@ public void addressGroupsEndpointShouldHandleRequests() throws Exception {
         .then()
             .statusCode(200)
             .contentType(JSON_CONTENT_TYPE)
-            .body(is("[\"user1@domain.com\", \"user2@domain.com\"]"));
+            .body(is("[\"user1@domain.com\",\"user2@domain.com\"]"));
```
Comment (mbaechler) [inline server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: so the test is bad because we should not care about json formating

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+    }
+
+    private void ensureNotShadowingAnotherAddress(MailAddress groupAddress) throws UsersRepositoryException {
+        if (usersRepository.contains(groupAddress.asString())) {
+            throw halt(HttpStatus.CONFLICT_409);
+        }
+    }
+
+
+    @PUT
```
Comment (aduprat) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java]: @DELETE

File: src/site/markdown/server/manage-webadmin.md [hunk 18]
```diff
@@ -425,3 +425,72 @@ Note that several calls to this endpoint will be run in a sequential pattern.
...
+Will return the group members as a list of JSON Strings representing mail addresses. For instance:
+
+```
+["member1@domain.com", "member2@domain.com"]
+```
+
+Response codes:
+ - 200: Success
+ - 400: Group structure is not valid
+ - 403: Server does not own the requested domain
```
Comment (aduprat) [inline src/site/markdown/server/manage-webadmin.md]: It seems that it's not what swagger documentation is saying: 403 & 409 are not thrown, IMHO.

File: server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java (hunk 16 above)
Comment (mbaechler) [reply server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/WebAdminServerIntegrationTest.java]: ok

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java [hunk 19]
```diff
@@ -0,0 +1,212 @@
...
+        @ApiImplicitParam(required = true, dataType = "string", name = USER_ADDRESS, paramType = "path")
+    })
+    @ApiResponses(value = {
+        @ApiResponse(code = 200, message = "OK", response = List.class),
+        @ApiResponse(code = 400, message = GROUP_ADDRESS + " or group structure format is not valid"),
+        @ApiResponse(code = 403, message = "server doesn't own the domain"),
+        @ApiResponse(code = 409, message = "requested group address is already used for another purpose"),
+        @ApiResponse(code = 500, message = "Internal server error - Something went bad on the server side.")
+    })
+    public HaltException addToGroup(Request request, Response response) throws org.apache.james.webadmin.utils.JsonExtractException, AddressException, RecipientRewriteTableException, UsersRepositoryException, DomainListException {
```
[the remark this answers is not in the document]
Comment (mbaechler) [reply server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java]: ok

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java (hunk 17 above)
Comment (mbaechler) [reply server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java]: ok

File: src/site/markdown/server/manage-webadmin.md (hunk 18 above)
Comment (mbaechler) [reply src/site/markdown/server/manage-webadmin.md]: have a look at RoutesGroups.addToGroup

Comment (aduprat) [reply src/site/markdown/server/manage-webadmin.md]: Yes, but in this documentation the `Response codes` seems to be for listing members of a group section.

**Comment by aduprat** [comment]
test this please

**Comment by aduprat** [comment]
test this please
